src/leiningen/ring/uberwar.clj

```diff
@@ -13,25 +17,54 @@
     (get-cp project)
     (->> (:library-path project) io/file .listFiles (map str))))
 
-(defn contains-entry? [^java.io.File file ^String entry]
-  (with-open [jar-file (JarFile. file)]
-    (some (partial = entry)
-          (map #(.getName ^JarEntry %)
-               (enumeration-seq (.entries jar-file))))))
-
 (defn jar-dependencies [project]
   (for [pathname (get-classpath project)
-        :let [file (io/file pathname)
-              fname (.getName file)]
-        :when (and (.endsWith fname ".jar")
-                   ;; Servlet container will have it's own servlet-api impl
-                   (not (contains-entry? file "javax/servlet/Servlet.class")))]
-    file))
-
-(defn jar-entries [war project]
+        :when (str/ends-with? pathname ".jar")]
+    (io/file pathname)))
+
+(def ^:private javax-servlet? #{"javax/servlet/Servlet.class"})
+
+(defn- pom-properties? [file-name]
+  (str/ends-with? file-name "pom.properties"))
+
+(defn- read-jar-pom-properties [^String full-path]
+  (with-open [in (.openStream (URL. full-path))]
+    (->> (doto (Properties.) (.load in))
+         (into {}))))
+
+(defn- from-pom-properties
+  "Returns a jar name which includes the groupId, artifactId, & version,
+   as read by the pom.properties file inside the jar file found at <jar-path>."
+  [jar-path pom-properties-jar-path]
+  (let [pom-properties-full-path (str "jar:file:" jar-path "!/" pom-properties-jar-path)
+        {:strs [groupId artifactId version]} (read-jar-pom-properties pom-properties-full-path)]
+    (when (and groupId artifactId version)
+      ;; found pom.properties - reconstruct the jar name to avoid collisions (issue #216)
+      (str groupId \- artifactId \- version ".jar"))))
+
+(defn- war-path-for-jar [^File jar]
+  (with-open [jar-file (JarFile. jar)]
+    (let [javax&pom (->> (.entries jar-file)
+                         enumeration-seq
+                         (map (fn [^JarEntry je] (.getName je)))
+                         (filter (some-fn javax-servlet? pom-properties?)))] ;; 2 elements maximum
+      ;; Servlet container will have it's own servlet-api impl
+      (when-not (some javax-servlet? javax&pom)
+        ;; we've confirmed that one of the two possible files is missing,
+        ;; so if we find one (via `first`), it will be the pom.properties
+        (str "WEB-INF/lib/"
+             (or (some->> (first javax&pom)
+                          (from-pom-properties (.getAbsolutePath jar)))
+                 ;; fallback to using the original jar name (best-effort approach)
+                 (.getName jar)))))))
+
+(defn- jar-entries
+  "Populates the 'WEB-INF/lib/' directory of the given <war> with this project's dependencies (jars).
```

Comment:
This docstring should be removed, and a better name for the function should be given. Perhaps something like `populate-war-with-dependent-jars`.